fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/UnnestedRecordTypeTest.java

```diff
@@ -0,0 +1,841 @@
+            typeBuilder.addParentConstituent(PARENT_CONSTITUENT, metaDataBuilder.getRecordType(OUTER));
+            typeBuilder.addNestedConstituent("middle", TestRecordsDoubleNestedProto.OuterRecord.MiddleRecord.getDescriptor(),
+                    PARENT_CONSTITUENT, field("many_middle", FanType.FanOut));
+            typeBuilder.addNestedConstituent("inner", TestRecordsDoubleNestedProto.OuterRecord.MiddleRecord.InnerRecord.getDescriptor(),
+                    "middle", field("inner", FanType.FanOut));
+            typeBuilder.addNestedConstituent("outer_inner", TestRecordsDoubleNestedProto.OuterRecord.MiddleRecord.InnerRecord.getDescriptor(),
+                    PARENT_CONSTITUENT, field("inner", FanType.FanOut));
+        };
+    }
+
+    @Nonnull
+    private static RecordMetaDataHook addKeyOtherIntValueIndex() {
+        return metaDataBuilder -> {
+            final KeyExpression expr = concat(
+                    field("map_entry").nest("key"),
+                    field(PARENT_CONSTITUENT).nest("other_id"),
+                    field("map_entry").nest("int_value")
+            );
+            metaDataBuilder.addIndex(UNNESTED_MAP, new Index(KEY_OTHER_INT_VALUE_INDEX, expr));
+        };
+    }
+
+    @Nonnull
+    private static RecordMetaDataHook addKeyOneKeyTwoValueOneValueTwo() {
+        return metaDataBuilder -> {
+            final KeyExpression expr = new KeyWithValueExpression(concat(
+                    field("entry_one").nest("key"),
+                    field("entry_two").nest("key"),
+                    field("entry_one").nest("value"),
+                    field("entry_two").nest("int_value")
+            ), 2);
+            metaDataBuilder.addIndex(TWO_UNNESTED_MAPS, new Index(KEY_ONE_KEY_TWO_VALUE_ONE_VALUE_TWO_INDEX, expr));
+        };
+    }
+
+    @Nonnull
+    private static RecordMetaDataHook addInnerFooOuterBarInnerBarIndex() {
+        return metaDataBuilder -> metaDataBuilder.addIndex(DOUBLE_NESTED, new Index(INNER_FOO_OUTER_BAR_INNER_BAR_INDEX,
+                concat(field("inner").nest("foo"), field("outer_inner").nest("bar"), field("inner").nest("bar"))));
+    }
+
+    @Nonnull
+    private static TestRecordsNestedMapProto.OuterRecord sampleMapRecord() {
+        return TestRecordsNestedMapProto.OuterRecord.newBuilder()
+                .setRecId(1066)
+                .setOtherId(1)
+                .setMap(TestRecordsNestedMapProto.MapRecord.newBuilder()
+                        .addEntry(TestRecordsNestedMapProto.MapRecord.Entry.newBuilder().setKey("foo").setValue("bar").setIntValue(1))
+                        .addEntry(TestRecordsNestedMapProto.MapRecord.Entry.newBuilder().setKey("baz").setValue("qux").setIntValue(2))
+                        .addEntry(TestRecordsNestedMapProto.MapRecord.Entry.newBuilder().setKey("wid").setValue("get").setIntValue(3))
+                )
+                .build();
+    }
+
+    @Test
```

Comment:
Sonarcloud detected that you have a `@Test` that will be silently ignored
```suggestion
    @Nonnull
```